### PR TITLE
feat: get stream based on bitrate and listeners

### LINKED ISSUE #38: automatic stream switching
With the new azuracast updates, relays have bitrate/format, unique identifiers, and per-stream listener counts.

the main strategy would be to directing listeners towards the relays.

here is the suggested logic

if relays exist choose a fast relays with least listeners.
if relays exist and speed is low, choose a low bitrate relay with least listeners.

if no relays choose fast non-relay
if relays exist and speed is low, choose low bitrate non-relay.

error handling on loading:
- If > 220, use the other streams
- all relay non-relay high bitrate streams 403ing? Drop the new user to 64
- On 403 during playback? Use another stream.

here are some of the issues I experienced this week:

GET https://coderadio-relay.freecodecamp.org/radio/8010/radio.mp3 net::ERR_INCOMPLETE_CHUNKED_ENCODING 200 (OK)

Failed to load resource: net::ERR_INCOMPLETE_CHUNKED_ENCODING

### PR DESCRIPTION
**Pull request by huyenltnguyen**
- [x] Get stream based on bitrate and number of listeners
- [x] Switch stream if there is an error

Closes #38

### CODE REVIEW

**Comment by ahmadabdolsaheb** [comment]
@huyenltnguyen any progress on this solution?

**Comment by huyenltnguyen** [comment]
@ahmadabdolsaheb sorry I haven't made any progress on the error handling. Reading the Nchan doc (https://nchan.io/) requires more time (and brains) than I thought, and I haven't been able to invest enough time on this.

Do you think it's okay to have this PR merged first, to enable the auto stream switching, and we will tackle the error handling in a future PR?

**Comment by ahmadabdolsaheb** [comment]
No worries.
Sure, I will give this a proper review before the weekend and merge if all is well.
Please feel free to work in parallel on error handling so you don't have to wait for me.

**Comment by ahmadabdolsaheb** [comment]
@huyenltnguyen it seems like we have an outage with remote streams, it might be an ideal time to test error handling.

File: src/components/Footer.js [hunk 1]
```diff
@@ -61,8 +61,8 @@ export default class Footer extends React.Component {
       mountOptions = (
         <select
           data-meta="stream-select"
-          defaultValue={this.props.url}
           onChange={this.handleChange.bind(this)}
+          value={this.props.url}
```
Comment (huyenltnguyen) [inline src/components/Footer.js:65]: This change is to make the dropdown renders the correct selected url.

**Comment by ahmadabdolsaheb** [comment]
Thanks for the update.
I am glad to hear you were able to resolve the error handling.
The interesting things is that if you go to https://coderadio-admin.freecodecamp.org/api/nowplaying
and click on the urls for the mounts in firefox after accepting the security alert, you can play it on the browser :). we need to figure out what is causing that.

**Comment by ahmadabdolsaheb** [comment]
I think the ssl certificate needs updating.

**Comment by ahmadabdolsaheb** [comment]
The issue with the certificate and the relays has been resolved.

**Comment by ahmadabdolsaheb** [comment]
The logic is great.
 I am just thinking of situations where this would cause issues on the main repo.
Let say if coderadio goes down, the playerError could go to an infinite loop and crash the platform.
Should we add a logic so when it tests all existing streams it would stop resetting the url?

**Comment by huyenltnguyen** [comment]
Yeah to be honest, I personally don't like the solution, since I fell into infinite a lot when I tested this, but I decided to pushed it up anyway to it is still a progress 😅 

I'll try to find a way to improve this 😄

**Comment by ahmadabdolsaheb** [comment]
we could have a state that keeps the streams  with the errors, so each time something errors out we only select one of the streams that has not been already in use.

make sure to change the state first and then change url. (I think you can pass a function that you would like to be called after state change as a parameter to setState).

**Comment by huyenltnguyen** [comment]
@ahmadabdolsaheb This PR is ready for another look (I pushed the changes yesterday but forgot to notify you 😅)

**Comment by ahmadabdolsaheb** [comment]
Thank you for your hard work.

**Pull request merged**